js/packages/teams-ai/src/authentication/MessagingExtensionAuthentication.ts

```diff
@@ -0,0 +1,111 @@
+import { ActivityTypes, InvokeResponse, TokenResponse, TurnContext } from 'botbuilder';
+import { FETCH_TASK_INVOKE_NAME, QUERY_INVOKE_NAME, QUERY_LINK_INVOKE_NAME } from '../MessageExtensions';
+import * as UserTokenAccess from './UserTokenAccess';
+import { OAuthPromptSettings } from 'botbuilder-dialogs';
+
+/**
+ * @internal
+ */
+export class MessagingExtensionAuthentication {
+    public async authenticate(context: TurnContext, settings: OAuthPromptSettings): Promise<string | undefined> {
+        const authObj = context.activity.value.authentication;
+
+        // Token Exchange
+        if (authObj && authObj.token) {
+            // Message extension token exchange invoke activity
+            const isTokenExchangable = await this.isTokenExchangeable(context, settings);
+            if (!isTokenExchangable) {
+                await context.sendActivity({
+                    value: { status: 412 } as InvokeResponse,
+                    type: ActivityTypes.InvokeResponse
+                });
+
+                return undefined;
+            } else {
+                const tokenExchangeResponse = await this.exchangeToken(context, settings);
```

Comment:
minor: the `isTokenExchangeable(...)` function already did a token exchange, may be able to optimize the code to not do token exchange again.
I'm doing the optimization in my branch, so no action should be required from your side. But if you have any comments regarding the optimization I'm doing, please let me know.

Reply:
yes please refactor this if it makes sense.